### PR TITLE
Autofocus effect

### PR DESCRIPTION
**Pull request by abernier**
Based on @ektogamat [work](https://twitter.com/Andersonmancini/status/1646837453053992960?s=20)  and pmndrs/postprocessing [DepthOfFieldEffect](https://pmndrs.github.io/postprocessing/public/docs/class/src/effects/DepthOfFieldEffect.js~DepthOfFieldEffect.html), `<Autofocus>` effect extends `<DepthOfField>` and drives its `target`.

https://user-images.githubusercontent.com/76580/235349072-e2c94588-7cf0-4311-a365-9856d80f3287.mov

Demo here: https://codesandbox.io/s/dfw6w4?file=/src/Autofocus.tsx

- [x] Documentation updated
- [x] Ready to merge

### CODE REVIEW

**Comment by ektogamat** [comment]
Great work my friend. Happy that this will become official 😊✌🏻

File: src/Autofocus.tsx [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+        dofRef,
+        hitpoint,
+        update,
+      }),
+      [hitpoint, update]
+    )
+
+    return (
+      <>
+        {debug &&
```
Comment (talentlessguy) [inline src/Autofocus.tsx:140]: could you please replace with `debug ? <...> : null`?

**Comment by talentlessguy** [comment]
Could you please also add an entry to the `docs/effects` folder?

**Comment by talentlessguy** [comment]
Could you please add a page to `docs/effects` instead of README?

File: src/Autofocus.tsx (hunk 1 above)
Comment (abernier) [reply src/Autofocus.tsx]: sure, great catch! thank you

**Review by talentlessguy — APPROVED** [review]

File: src/Autofocus.tsx [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+        } else {
+          const { x, y } = followMouse ? pointer : { x: 0, y: 0 }
+          const hit = await getHit(x, y)
+          if (hit) hitpoint.copy(hit)
+        }
+
+        // Update target
+        if (updateTarget && dofRef.current?.target) {
+          if (smoothTime > 0 && delta > 0) {
+            easing.damp3(dofRef.current.target, hitpoint, smoothTime, delta)
```
Comment (abernier) [inline src/Autofocus.tsx:104]: @drcmda damping target is here directly supported by the component, but as mentioned in the doc, the ref-api allows doing it from the outside.: https://github.com/pmndrs/react-postprocessing/blob/6f733b2a469c79bf8de2bb2b877455e256a476d5/docs/effects/autofocus.mdx?plain=1#L40-L48

If you prefer, I can remove the `maath` dep and remove damping feature, but I think it is handy/user-friendly to have it by default.

Let me know ;)

**Comment by ektogamat** [comment]
Can I help with something? Documentation maybe? I have no idea how to contribute hehehe. Just want to be useful if needed, okay? 🫶🏻

File: src/Autofocus.tsx (hunk 2 above)
Comment (drcmda) [reply src/Autofocus.tsx:103]: maath is good imo

File: src/Autofocus.tsx [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+        return hit ? ndc.unproject(camera) : false
+      },
+      [ndc, depthPickingPass, camera]
+    )
+
+    const [pointer] = useState(new THREE.Vector2())
+    useEffect(() => {
+      if (!followMouse) return
+
+      async function onPointermove(e: PointerEvent) {
```
Comment (drcmda) [inline src/Autofocus.tsx:74]: not needed imo, useFrame has that ootb: state.pointer

File: src/Autofocus.tsx [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+      },
+      [target, hitpoint, followMouse, getHit, smoothTime, pointer]
+    )
+
+    useFrame(async (_, delta) => {
+      if (manual) return
+      update(delta)
+    })
+
+    useFrame(() => {
```
Comment (drcmda) [inline src/Autofocus.tsx:117]: better combine into one useFrame, one function invocation less

**Review by drcmda — CHANGES_REQUESTED** [review]

File: src/Autofocus.tsx (hunk 3 above)
Comment (abernier) [reply src/Autofocus.tsx]: omg yes, totally missed that, it a great simplification! thank you

**Pull request merged**